crates/sui-metrics-proxy/src/admin.rs

```diff
@@ -0,0 +1,160 @@
+use anyhow::{Context, Result};
+use axum::{routing::post as axum_post, Router};
+use fastcrypto::ed25519::Ed25519KeyPair;
+use fastcrypto::ed25519::Ed25519PublicKey;
+use fastcrypto::traits::KeyPair;
+use fastcrypto::traits::ToFromBytes;
+use serde::Deserialize;
+use std::sync::Arc;
+use std::time::Duration;
+use sui_tls::{
+    rustls::ServerConfig, SelfSignedCertificate, TlsAcceptor, ValidatorAllowlist,
+    ValidatorCertVerifier,
+};
+use sui_types::sui_system_state::ValidatorMetadata;
+use tokio::{signal, sync::mpsc};
+use tracing::{error, info};
+
+/// Configure our graceful shutdown scenarios
+pub async fn shutdown_signal(h: axum_server::Handle) {
+    let ctrl_c = async {
+        signal::ctrl_c()
+            .await
+            .expect("failed to install Ctrl+C handler");
+    };
+
+    #[cfg(unix)]
+    let terminate = async {
+        signal::unix::signal(signal::unix::SignalKind::terminate())
+            .expect("failed to install signal handler")
+            .recv()
+            .await;
+    };
+
+    #[cfg(not(unix))]
+    let terminate = std::future::pending::<()>();
+
+    tokio::select! {
+        _ = ctrl_c => {},
+        _ = terminate => {},
+    }
+
+    let grace = 30;
+    info!(
+        "signal received, starting graceful shutdown, grace period {} seconds, if needed",
+        &grace
+    );
+    h.graceful_shutdown(Some(Duration::from_secs(grace)))
+}
+
+/// App will configure our routes and create our mpsc channels.  This fn is also used to instrument
+/// our tests
+pub fn app(buffer_size: usize) -> Router {
+    // we accept data on our UpstreamConsumer up to our buffer size.
+    let (sender, receiver) = mpsc::channel(buffer_size);
+    let mut consumer = UpstreamConsumer::new(receiver);
+
+    tokio::spawn(async move { consumer.run().await });
+
+    // build our application with a route and our sender mpsc
+    Router::new()
+        .route("/publish/metrics", axum_post(publish_metrics))
+        .with_state(Arc::new(sender))
+}
+
+pub async fn server(
+    listener: std::net::TcpListener,
+    acceptor: TlsAcceptor,
+    app: Router,
+) -> std::io::Result<()> {
+    // setup our graceful shutdown
+    let handle = axum_server::Handle::new();
+    // Spawn a task to gracefully shutdown server.
+    tokio::spawn(shutdown_signal(handle.clone()));
+
+    axum_server::Server::from_tcp(listener)
+        .acceptor(acceptor)
+        .handle(handle)
+        .serve(app.into_make_service())
+        .await
+}
+
+pub fn create_server_cert(
+    hostname: &str,
+) -> Result<(ServerConfig, ValidatorAllowlist), sui_tls::rustls::Error> {
+    let mut rng = rand::thread_rng();
+    let server_keypair = Ed25519KeyPair::generate(&mut rng);
+    let server_certificate = SelfSignedCertificate::new(server_keypair.private(), hostname);
+
+    ValidatorCertVerifier::rustls_server_config(
+        vec![server_certificate.rustls_certificate()],
+        server_certificate.rustls_private_key(),
+    )
+}
```

Comment:
This isn't going to work with how the client is configured today. Today the client side is expecting the server's cert to be a valid cert from a well known CA, not a self-signed on. for testing purposes you may need to tweak how the http client is constructed to be able to ignore invalid server certs.

Reply:
You'll have to use https://docs.rs/reqwest/latest/reqwest/struct.ClientBuilder.html#method.danger_accept_invalid_certs

here: https://github.com/MystenLabs/sui/blob/main/crates/sui-node/src/metrics.rs#L72

for testing, but we wont want to commit that. I suppose we could add an escape hatch via a config to configure ignoring server certs here....

Reply:
deferring tls until the next pr